oeps/oep-0052/decisions/001-schema-representation.rst

```diff
@@ -0,0 +1,35 @@
+Don't use openedx-events Attrs Classes to Represent Deserialized Event Bus Objects
+==================================================================================
+
+Context
+-------
+
+The event bus client libraries have their own object representation of messages. In pulsar, they are defined as `Record`
+objects. These objects map easily to Avro Schema definitions and can be used to define the schema. The attrs objects
+that we currently have for signals based events don't easily inter-operate with event bus client objects.
+
+Decision
+--------
+
+We will define and manage event bus schema definition and related object definition independently of the signals based
+attrs objects.
+
+Consequences
+------------
+
+* We will have potentially different schema for what data is available within a service compared to between services.
+
+* The initial implementation will require a dependency on pulsar/kafka in the shared library that holds the relevant
+  schema and objects.
+
+Rejected Alternatives
+---------------------
+
+Make new schema classes that can generate schema based on the `attrs` classes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It should be possible to use the introspection information that attrs provides for
+[extension](https://www.attrs.org/en/stable/extending.html) to be able to generate schema based on attrs classes. This
+along with some guidelines could allow us to use the attrs classes as the de-facto definition of schema for the Open edX
+Platform. We are currently rejecting this approach because of the level of effort involved. We don't want to wait for
+this to exist before we can benefit from using the event bus. In the future we may still want to build this.
```

Comment:
How complex are these schema classes? The ability to more or less seamlessly translate from in-process events to cross-network events was a major design goal for openedx-events, and I'm worried that not having a translation layer like this will either lock us into a particular vendor (despite our best intentions) or else cause feature drift between the definitions of those data structures and their message bus equivalents.

Reply:
The big issue with using our existing event objects vs using ones based on AVRO schema is deciding which is the source of truth.  With both Pulsar and Kafka, we need to be able to pass in the AVRO schema and a class that can serialize/deserialize it.

Long term, I think we could generate that schema from our Attrs classes but we didn't want solving that to block moving forward with getting something working.  I think it makes sense to build that in before cross service messaging is "fully available" though.

Reply:
@feanil: Maybe we could replace:
```
In the future we may still want to build this.
```
with something more definitive, and tied to your last statement, like:
```
We hope to build this before external events are required as part of the Open edX platform, but we are not committing to this at this time.
```

Reply:
@ormsbee: We are also still interested in pursuing the idea of an abstraction layer, and seeing how that might be possible, which would clearly affect the final design/implementation.

Reply:
Schema generation from Attrs class is now implemented at https://github.com/eduNEXT/openedx-events/blob/main/openedx_events/avro_attrs_bridge.py

Reply:
I'm resolving this to make reviewing and finishing this PR easier. Feel free to unresolve if you want to discuss further.